proposals/ref-unsafe-in-iterators-async.md

```diff
@@ -43,11 +43,98 @@ lock (this)
 The following changes are tied to LangVersion, i.e., C# 12 and lower will continue to disallow
 ref-like locals and `unsafe` blocks in async methods and iterators,
 and C# 13 will lift these restrictions as described below.
+However, spec clarifications which match the existing Roslyn implementation should hold across all LangVersions.
 
 [§13.3.1 Blocks > General][blocks-general]:
 
-> ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).~~
-> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.
+> A *block* that contains one or more `yield` statements ([§13.15][yield-statement]) is called an iterator block,
+> **even if those `yield` statements are contained only indirectly in nested blocks (excluding nested lambdas and local functions).**
+>
+> [...]
+>
+> ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).
+> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.~~
+> **The iterator block used to implement an iterator ([§15.14][iterators])
```

Comment:
> **The iterator block used to implement an iterator ([§15.14][iterators]) [](http://example.com/codeflow?start=1&length=73)

Wouldn't it be more intuitive to treat the entire iterator declaration as defining a safe context, since presence of `unsafe` modifier makes a difference? Was this option seriously considered?

Reply:
> Wouldn't it be more intuitive to treat the entire iterator declaration as defining a safe context

That might be a more straightforward alternative, thanks. The way I thought about this was that an iterator is defined by the block (method body or property accessor) which contains `yield`, so that's where the safe context should begin (and hence the signature and the `set` accessor are not in that context).

Reply:
Will it be possible to implement that though? Do we know whether a method is an iterator when binding its signature?

Reply:
>  Also I'm not sure if partial declarations without bodies can "see" the corresponding partial implementation body to determine whether it's an iterator.

I think partial declarations are never iterators. They lack an iterator block